assets/src/edit-story/components/library/panes/shared/pillGroup/pillGroup.js

```diff
@@ -0,0 +1,195 @@
+import PropTypes from 'prop-types';
+import { rgba } from 'polished';
+import { v4 as uuidv4 } from 'uuid';
+
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import { ArrowDown } from '../../../../button';
+import { useKeyDownEffect } from '../../../../keyboard';
+import useRovingTabIndex from '../useRovingTabIndex';
+import Pill from './pill';
+
+/**
+ * Internal dependencies
+ */
+import {
+  PILL_COLLAPSED_FULL_HEIGHT,
+  PILL_BOTTOM_MARGIN,
+  PILL_TOP_MARGIN,
+} from './constants';
+import useExpandAnimation from './useExpandAnimation';
+import useHandleRowVisibility from './useHandleRowVisibility';
+
+const Section = styled.div`
+  height: ${PILL_COLLAPSED_FULL_HEIGHT}px;
+  min-height: ${PILL_COLLAPSED_FULL_HEIGHT}px;
+  background-color: ${({ theme }) => rgba(theme.colors.bg.workspace, 0.8)};
+  display: flex;
+  flex-direction: column;
+  align-items: flex-start;
+  flex: 0 1 auto;
+  position: relative;
+  transition: height 0.2s, min-height 0.2s;
+`;
+
+// This hides the pills unless expanded
+const Container = styled.div`
+  overflow: hidden;
+  margin: ${PILL_TOP_MARGIN}px 12px ${PILL_BOTTOM_MARGIN}px 24px;
+`;
+
+const InnerContainer = styled.div`
+  display: flex;
+  justify-content: center;
+  flex-wrap: wrap;
+  position: relative;
+  transition: transform 0.2s;
+`;
+
+// Flips the button upside down when expanded;
+// Important: the visibily is 'inherit' when props.visible because otherwise
+// it gets shown even when the provider is not the selectedProvider!
+const ExpandButton = styled(ArrowDown)`
+  display: flex;
+  position: absolute;
+  bottom: -16px;
+  background: ${({ theme }) => theme.colors.fg.gray16};
+  max-height: none;
+  width: 32px;
+  height: 32px;
+  border-radius: 16px;
+  ${({ isExpanded }) => isExpanded && 'transform: matrix(1, 0, 0, -1, 0, 0);'}
+  visibility: inherit;
+  align-self: center;
+  justify-content: center;
+  align-items: center;
+`;
+
+const PillGroup = ({ items, selectedItemId, selectItem, deselectItem }) => {
+  const [isExpanded, setIsExpanded] = useState(false);
+
+  const sectionRef = useRef();
+  const innerContainerRef = useRef();
+
+  const itemRefs = useRef([]);
+
+  const [focusedRowOffset, setFocusedRowOffset] = useState(0);
+
+  const handleClick = (selected, id) => {
+    if (selected) {
+      deselectItem();
+    } else {
+      setIsExpanded(false);
+      selectItem(id);
+    }
+  };
+
+  useExpandAnimation({
+    sectionRef,
+    innerContainerRef,
+    isExpanded,
+    setFocusedRowOffset,
+  });
+
+  useHandleRowVisibility({
+    isExpanded,
+    innerContainerRef,
+    selectedItemId,
+    setFocusedRowOffset,
+    itemRefs,
+  });
+
+  const hasItems = items.length > 0;
+  useRovingTabIndex({ ref: sectionRef }, [isExpanded]);
+  useKeyDownEffect(
+    sectionRef,
+    !isExpanded ? 'down' : [],
+    () => hasItems && setIsExpanded(true),
+    [isExpanded, hasItems]
+  );
```

Comment:
One tiny nit: You can use the left arrow to move to an invisible option while the list is collapsed (if you have selected an item further down that the first item). Maybe this one should just expand on any arrow press and not only on down arrow?

Reply:
Hmm, I'm not sure -- in this case, ideally, it should do the moving + expand at the same time, otherwise, if the user is on the first item and clicks `right`, it would just expand but not move. Just tested it out and it feels a bit unexpected.

I'll merge this as-is and we can consider/test it again when implementing the text sets panel.

Reply:
That's fine - it's a tiny edge case.